enhancements/machine-api/machine-pools-in-cluster.md

```diff
@@ -0,0 +1,240 @@
+
+## Summary
+
+Today we expose a useful construct in the Installer's InstallConfig known as MachinePool. When used the installer will query the available AZs for the account, and generate the relevant MachineSets to spread the requested number of machine replicas out over the AZs.
+
+This enhancement proposes bringing this functionality into the cluster itself for use by customers on an on-going basis.
+
+
+## Motivation
+
+The Installer MachinePool is a useful construct that is only available at install time. Post-install customers must manage this spread themselves.
+
+[OpenShift Hive](https://github.com/openshift/hive) also supports day 2 MachinePools using the same code from the Installer. However because Hive must install and maintain MachineSets for many versions of OCP, this code re-use between the Installer has generated a number of difficult and messy bugs when things change in the Installer. Bringing this feature in-cluster would also give Hive a stable API to interact with instead of fighting with one version of vendored Go code to be used across many OCP versions and a significant number of logic forks based on OCP version.
+
+### Goals
+
+List the specific goals of the proposal. How will we know that this has succeeded?
+
+  * Provide customers with an API to automatically spread machines out over the AZs in their account.
+  * Provide OpenShift Hive with a stable API to manage MachineSets spread over AZs day 2.
+
+### Non-Goals
+
+  * No linkup with the MachineDeployment concept for rolling updates of Machines.
+
+## Proposal
+
+### User Stories [optional]
+
+#### Story 1
+
+As an OCP user, I would like to be able to specify min/max machine replicas and
+have those machines automatically spread out amongst all Availability Zones
+that are available in my account.
+
+#### Story 2
+
+As an OpenShift Hive developer, I would like a stable API for spreading
+machines across Availability Zones. This would allow me to sync an API
+in-cluster rather than use one version of vendored code from the installer to
+generate each MachineSet to be reconciled.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Implementation would require a new CRD and controller aligned hopefully with a pre-existing operator in OCP.
+
+The installer MachinePool definition can be seen [here](https://github.com/openshift/installer/blob/master/pkg/types/machinepools.go).
+
+The code for spreading MachineSets across AZs [already exists today](https://github.com/openshift/installer/blob/master/pkg/asset/machines/aws/machinesets.go) and could be removed from the installer, which would then just pass through an instance of the new API.
+
+The [Hive MachinePool API](https://github.com/openshift/hive/blob/master/pkg/apis/hive/v1/machinepool_types.go) also includes support for enabling auto-scaling on the resulting MachineSets, and this functionality should be carried over to the new API even if not initially supported by the Installer.
```

Comment:
This seems like a tangential issue to the main aim of the proposal, do we want to include this here or can we drop this for now and narrow the scope to just the multi-az support?

Reply:
Somewhat depends on what the outcome is here. If we end up with upstream cluster-api MachineSets supporting multiple AZs, Hive likely then just has to sync one stable MachineSet  and one stable Autoscaler config, which seems ok.

However if that is not the case and we do end up with a new MachinePool API of some form, then Hive would be syncing a singular MachinePool but then would need to introspect what was done to sync a per MachineSet autoscaling config. If we've got a higher level construct in-cluster, IMO that needs to encompass the autoscaling config to apply to what it generates.